### DISCUSSION TITLE
Third Party Hooks

### DISCUSSION THREAD

**Discussion by dgduncan** [mochi-mqtt/server#235] [Ideas]
Hey @mochi-co I would love for you if ya have a second to check out [this](https://github.com/dgduncan/mochi-cloud-hooks). It's a hooks library I have been working on in my spare time for this broker. I currently use all of the hooks that I have there as each one has been written to resolve my own use case. No need to link them but would love to get your feedback and idea if this is something that you would be interested in me pursuing further. I'll be the first to admit I haven't kept the repo as "clean" and "OSS" ready as I would usually as this has been very hobby on the side but would still love the feedback.

**Comment by mochi-co** [comment]
@dgduncan This is amazing! I'm impressed! 

Some simple thoughts:

- The code is clean and readable, I love that you have proper tests. 
- I'd love to see some examples of how to use the different hooks and in what circumstances/how you are using them.
- Given that it's focused on GCP I would specify that in the repo name - e.g. mochi-gcp-hooks.

My only note of concern - I would be cautious about using GPL3 license as it creates a massive legal headache for anyone looking to use hooks in a non-open or commercial codebase. Many companies will outright prohibit inclusion of GPL code because it can potentially require them to open source their entire proprietary codebase. This is one of the reasons mochi-co/mqtt is MIT license. I totally understand if this is a deliberate choice, but if it's not and you're just looking to ensure attribution, Apache is a pretty solid alternative. https://brainhub.eu/library/open-source-licenses-to-avoid has a good list of the pros and cons of each license. 

Overall it's very promising and I'd be thrilled to link it in the README if you want (even as 'experimental')!
